Add restricted content blocks

gkaragia opened the pull request:
### Changes proposed in this Pull Request

* This PR introduces a new blocks to the single course page, 'Restrict Content' (naming suggestions welcome 😄 ).
* The block is similar to how the 'Group' block works but they show the content to enrolled, unenrolled or users which completed a course only.
* I investigated to possibly add the blocks to the single lesson page too. Currently in the single lesson template we don't show the content at all when the user is not enrolled. For this reason we should probably wait until we drop the single lesson template to see if it fits to add these blocks.

_Update:_
* It is possible to transform any selection of blocks to move it inside a wrapper 'Enrolled Content' or 'Unenrolled Content' block. This allows users easily to restrict the content after it is created. When transforming one of these blocks to the inverse one, a wrapper is not created and the block is switched instead.
![record](https://user-images.githubusercontent.com/53191348/102106933-d571e000-3e39-11eb-8fc1-d8fc40a94e43.gif)

_Second Update:_
* There is now a single block for all cases. The user can select between different options by using a dropdown. It also has a new option which restricts the content to users which completed a course.
![button](https://user-images.githubusercontent.com/53191348/104202081-22de6e80-5433-11eb-93fa-2ffc09b607ef.gif)
* Since there wasn't an intuitive option to remove the restriction once the block is created, I also added an option to the drop down menu which removes the 'Restrict content' block. Basically it replaces the block whith its inner blocks
![remove](https://user-images.githubusercontent.com/53191348/104202300-6a64fa80-5433-11eb-87e9-dfa6b9fc26e3.gif)
* I have considered the suggestions in [this comment](https://github.com/Automattic/sensei/pull/3849#issuecomment-748191152) but I decided against them for the following reasons:
  * The toolbar should follow gutenberg's minimalistic approach so any not absolutely necessary text should be avoided. The ideal way to do this would be to have the selected icon on the toolbar and the options on the dropdown should be a button with the icon and a few words, similar with the rest of the dropdown menus. Additionally this would require more time to implement.
  * I used text only for now to avoid the extra work of creating new icons.
  * Since the button can be considered as a title too, I think that we don't need any extra indicator to explain that the content will be hidden. We can revise this if users find the block confusing.


### Testing instructions

* In a course include the new block and add some content in them.
* Try the 'Enrolled Users', 'Unenrolled Users' and 'Course Completed' options and observe that the content is displayed (or not) accordingly in the frontend.
* Try selecting multiple blocks and transform them to a 'Restrict content' block.
* Try removing the restriction from a block.

gkaragia commented on includes/blocks/class-sensei-lesson-blocks.php:
```diff
@@ -0,0 +1,67 @@
+<?php
```
Written this class as part of my investigation of adding the blocks to the lesson page. Happy to remove it since it is unused.

donnapep commented:
Dropping a quick note that this one should go through a design review before being merged. Making a note to do that this week.

yscik commented:
Nice!

A few things that would be good to have in this:

* Add block transform to wrap any block in Enrolled/Unenrolled. This way the user can group existing blocks into a restricted block. This probably also covers switching Enrolled to Unenrolled and vice-versa. 
* Full/wide alignment options do not work inside these blocks. Maybe they could be always full-width? Or have alignment options, but could be cleaner if they are just layout-agnostic, if it's possible.

yscik commented:
Design-wise, I think the main thing here is a way to indicate in the editor when content is restricted. This is kind of a wider issue of making a parent block more visible, that also came up with the course outline block, and will hopefully be handled by Gutenberg at some point. Until then, we should roll our own solution, but it's worth a dedicated PR.

The core Gallery block already selects both the container an the inner block:

<img width="630" alt="image" src="https://user-images.githubusercontent.com/176949/102084190-14da0580-3e15-11eb-9835-3543844d6b04.png">

Jetpack's *Repeat Visitor* block has a note, but it's unclear where the block ends:

<img width="591" alt="image" src="https://user-images.githubusercontent.com/176949/102085133-823a6600-3e16-11eb-95d8-3e01b22d776b.png">

I also explored some options in Conditional Sections for Course Enrollment Blocks (p6rkRX-1Dd-p2):

Outline and labels:
<img width="638" alt="image" src="https://user-images.githubusercontent.com/176949/102086214-05a88700-3e18-11eb-8f02-72951948995b.png">

Sections:
<img width="777" alt="image" src="https://user-images.githubusercontent.com/176949/102084830-06d8b480-3e16-11eb-89d9-b4bdb35b7bb9.png">

alexsanford commented:
> Dropping a quick note that this one should go through a design review before being merged.

cc @pablohoneyhoney

gkaragia commented:
> Nice!
> 
> A few things that would be good to have in this:
> 
>     * Add block transform to wrap any block in Enrolled/Unenrolled. This way the user can group existing blocks into a restricted block. This probably also covers switching Enrolled to Unenrolled and vice-versa.
> 
>     * Full/wide alignment options do not work inside these blocks. Maybe they could be always full-width? Or have alignment options, but could be cleaner if they are just layout-agnostic, if it's possible.

Great ideas! Added transforms and alignment options (following the group block).

yscik commented on assets/blocks/restricted-content/index.js:
```diff
@@ -0,0 +1,115 @@
+			type: 'block',
+			isMultiBlock: true,
+			blocks: [ '*' ],
+			__experimentalConvert: ( blocks ) => {
```
There is a `transform(attributes, innerBlocks)` function as a stable API, any reason this one is used?

gkaragia replied:
Yes transform includes the attributes and inner blocks only but we need the plugin name in convert. It also supports better multiple selections. It also seems to be the plan to [deprecate transform in favor of convert](https://github.com/WordPress/gutenberg/pull/19401) so I thought that it is safe enough to use.

yscik commented on assets/blocks/restricted-content/enrolled/block.json:
```diff
@@ -0,0 +1,8 @@
+  "name": "sensei-lms/enrolled-content",
+  "category": "sensei-lms",
+  "supports": {
+    "html": false,
```
I think it's fine to allow html editing for these.

gkaragia replied:
We did some tests with it and there seems do be no useful use cases for this so we are going to keep it as it is.

yscik commented on includes/blocks/class-sensei-course-blocks.php:
```diff
@@ -42,21 +42,53 @@ class Sensei_Course_Blocks {
 		$this->progress        = new Sensei_Course_Progress_Block();
 		$this->contact_teacher = new Sensei_Block_Contact_Teacher();
 		$this->take_course     = new Sensei_Block_Take_Course();
+		new Sensei_Restricted_Content_Block();
```
Should we also store this in a variable, just to keep it accessible and consistent with the rest?

gkaragia replied:
I don't think that we should do that unless we specifically need this instance to be accessed from somewhere else. Even then it would be better to pass it as a dependency than define it as a public attribute.

yscik commented on includes/blocks/class-sensei-course-blocks.php:
```diff
@@ -86,7 +118,7 @@ public function enqueue_block_editor_assets() {
 		}
 
-		Sensei()->assets->enqueue( 'sensei-blocks', 'blocks/index.js', [], true );
+		Sensei()->assets->enqueue( 'sensei-blocks', 'blocks/course.js', [], true );
```
The name could also be updated here, for something like `sensei-single-course-blocks`.

gkaragia replied:
Done in 2dfdd8595c65724589ceeb63ff77231b577fd1b9

yscik commented on assets/blocks/restricted-content/index.js:
```diff
@@ -0,0 +1,115 @@
+};
+
+export const UnenrolledContent = {
+	title: __( 'Unenrolled Content', 'sensei-lms' ),
+	description: __(
+		'Content inside this container block will be displayed to unenrolled users only.',
```
I don't think we use `unenrolled` anywhere, something like `..users not enrolled` might sound better.

gkaragia replied:
Fixed in 94e3470fb49c41a0f64056b0bd2aa6bcd7859bcc. I don't like `unenrolled` too but couldn't find a better alternative. Hoping for a better recommendation by the design review cc @donnapep

yscik reviewed (commented):
Code is looking code, just a few notes:

renatho commented on assets/blocks/sensei-single-course-blocks.js.js:
```diff
@@ -7,6 +7,7 @@ import {
 	CourseOutlineLessonBlock,
```
This file name is with a double `.js` extension.

gkaragia replied:
Oops d3605de42fe8e583a34ae61137a9ffa6839a9589

renatho reviewed (commented):
Cool! I think this feature will be very useful! Did you also think about the option to create a single block with multiple states as a setting (enrolled, unenrolled, course completed...)?

For the design meeting, maybe would be nice to think about the unenrolled icon too. The open padlock gives a sensation that it is unlocked for any user (although the title says clearly the block purpose).

cc @donnapep

gkaragia commented:
> Cool! I think this feature will be very useful! Did you also think about the option to create a single block with multiple states as a setting (enrolled, unenrolled, course completed...)?

Yep I considered it but thought that two separate block is probably more easy to understand.

pablohoneyhoney commented:
While the use case makes sense, it'd be better to think of this as an implementation that shares paradigms to similar cases like Pay Wall in Dotcom. 

@alexsanford worth connecting with @fditrapani and @artpi 
Some context https://heytimmyneutron.wordpress.com/2020/03/04/premium-content-block-i3/ 
I haven't found related issues. 

Additionally, it's more natural to think of the premium/extra content as the block to be added, rather than creating a paradox of rolled/unrolled. So the creator has the clarity that in order to add special content, they need to add the Premium (or whatever the right term is) block.

alexsanford commented:
> shares paradigms to similar cases like Pay Wall in Dotcom.

Good point, the "Premium Content" block on dotcom seems like a good place from which to draw inspiration.

> Additionally, it's more natural to think of the premium/extra content as the block to be added, rather than creating a paradox of rolled/unrolled.

I think both paradigms are needed. A concrete example that came up when building WP Courses was the need to have sales copy on the Course page that only displays for non-enrolled users.

Note that the Premium Content block has this capability; the "Visitor View" is a fully editable container. Perhaps it would make sense to do something similar, having a "Visitor View" and "Learner View"?

yscik commented:
The content for Enrolled/unenrolled users is not necessarily connected — someone could just want to hide the outline to guest users, or add some extra content, like a sales copy, or a welcome message. I wouldn't push this in the direction where the content is toggled based on a state.

We can merge these into a single `Restricted` block, with a setting on who can see it:
<img width="604" alt="image" src="https://user-images.githubusercontent.com/176949/102636209-a4a0ec00-4154-11eb-947f-118f020cf8f8.png">

I think the main challenge here is to indicate the visibility of these blocks when they are not selected, so the user can have a better overview of what's in the post. Even just a subtle icon would help:
<img width="778" alt="image" src="https://user-images.githubusercontent.com/176949/102637724-b7b4bb80-4156-11eb-8cd2-80b4d1c55a00.png">

renatho commented on includes/blocks/class-sensei-course-outline-block.php:
```diff
@@ -57,41 +57,22 @@ class Sensei_Course_Outline_Block {
 
 	/**
-	 * Initialize block instance.
+	 * Resets block content.
```
Should we rename the method too to match the described purpose?

gkaragia replied:
I didn't do that because it is a public method and we should deprecate it first. But truth is that there is practically no chance anybody using it 1ebae7abff101d64f8dbf23373eed1d6c65dbeaf

renatho commented on assets/blocks/restricted-content/settings.js:
```diff
@@ -0,0 +1,140 @@
+ *
+ * @param {string} clientId The block client id.
+ */
+const useIsSingleRestrictSelected = function ( clientId ) {
```
Could we change these functions to arrow functions, just to keep the standard? :)

gkaragia replied:
254ed678535a1e3ac3beb108a9d70f2e42a77bbe

renatho replied:
I think we missed the `useOnRestrictionRemoval` too. :)

gkaragia replied:
Ah! Missed that 216adca3b33831d36867e2ed4175917b321e8fb4

renatho commented on assets/blocks/restricted-content/settings.js:
```diff
@@ -0,0 +1,140 @@
+									onClose();
+								} }
+							>
+								{ __( 'Remove restriction', 'sensei-lms' ) }
```
Cool! Nice idea!

renatho reviewed (approved):
It works well, and from the code perspective, it looks good to me!
I added a few comments, but nothing critical.

Before merging, it worths it to wait for UX/product feedback.

> I investigated to possibly add the blocks to the single lesson page too. Currently in the single lesson template we don't show the content at all when the user is not enrolled. For this reason we should probably wait until we drop the single lesson template to see if it fits to add these blocks.

Nice! When we can, I think it would be nice to use this block in the lessons too, and maybe in the quizzes.

yscik commented:
One minor thing, the toolbar button and dropdown should be left aligned, and have no extra spacing/fixed width. It's fine if the width changes when the user selects something else.
<img width="179" alt="image" src="https://user-images.githubusercontent.com/176949/104933942-d2728e00-59a9-11eb-8c53-cd9a4c2d3455.png">

gkaragia commented on assets/blocks/restricted-content/block.json:
```diff
@@ -0,0 +1,14 @@
+    "align": [ "wide", "full" ]
+  },
+  "attributes": {
+    "optionSelected": {
```
Done in 6f04c7a97876d2a65f69f4392dee94255d2a0000

alexsanford commented:
A few notes. These are **not** high-priority items, I'm fine if we address them a new PR (as long as they don't get lost).

### Adding the block

The user experience for adding the block is a bit weird. When I want to add restricted content, I see this:

<img width="520" alt="Screen Shot 2021-01-18 at 13 11 06" src="https://user-images.githubusercontent.com/842193/104945129-284a3480-59b0-11eb-9d15-fddb2f6a5c0c.png">

I don't have any indication that the Restricted Content block is related to Sensei or a user's status within a course. And it's not clear that the "Member Content" and "Non-Member Content" is related to a different plugin entirely. This is more clear in the sidebar (screenshot below) but not the main inserter.

<img width="352" alt="Screen Shot 2021-01-18 at 13 13 23" src="https://user-images.githubusercontent.com/842193/104945325-795a2880-59b0-11eb-8e05-c03f4b9116a9.png">

I wonder if calling it something like "Restricted Course Content" might be more clear?

Similarly, the description text doesn't give any indication that this is for Sensei, and that it relates to a user's status within a course.

<img width="295" alt="Screen Shot 2021-01-18 at 13 29 10" src="https://user-images.githubusercontent.com/842193/104946864-b0c9d480-59b2-11eb-907a-23557ac38412.png">


### Adding content to the block

Next, when I add the "Restricted Content" block, I expect to be able to start typing my content (or using a slash command to add a block). Instead I see this:

<img width="744" alt="Screen Shot 2021-01-18 at 13 19 41" src="https://user-images.githubusercontent.com/842193/104945996-5b40f800-59b1-11eb-9350-5a410b46e0d3.png">

I have to click the `+` button, and choose the "Paragraph" block before I can start typing. This seem like needless friction.

That being said, I'm not sure how to solve this, because if we add a paragraph by default, we will not be able to show the Restricted Content block's toolbar, so maybe this is the best we can do for now 🤷‍♂️ 

### Styling Issue on Twenty Twenty-One and Seedlet

On Twenty Twenty-One, when I add several paragraphs to the Restricted Content block, the styling on the frontend does not seem to work properly. Here is what I see in the editor:

<img width="747" alt="Screen Shot 2021-01-18 at 13 14 19" src="https://user-images.githubusercontent.com/842193/104945402-9abb1480-59b0-11eb-9735-7d5764af9056.png">

And here is the frontend:

<img width="696" alt="Screen Shot 2021-01-18 at 13 14 32" src="https://user-images.githubusercontent.com/842193/104945419-a3134f80-59b0-11eb-9154-8246ad89d994.png">

Note that the paragraphs inside the block do not have the right separation between them.

This also occurs on the Seedlet theme.

I've also tested on Twenty Twenty, Storefront, and GeneratePress. It seems to work fine on all of those.

gkaragia commented:
In general I have used the 'Group' block as an example for this one.

> I don't have any indication that the Restricted Content block is related to Sensei or a user's status within a course.
> I wonder if calling it something like "Restricted Course Content" might be more clear?

This is the downside of using a generic block for every case as it limits us to use a generic name. Maybe we should rename it to 'Restricted Sensei Content'? Don't like it as a name but can't find something better.

> Similarly, the description text doesn't give any indication that this is for Sensei, and that it relates to a user's status within a course.

Updated the description here: 5caa08f4d8aff48a7110c8e6a4c508e06e21c1c3

> Next, when I add the "Restricted Content" block, I expect to be able to start typing my content (or using a slash command to add a block). Instead I see this:

Followed the design of the Group block. I tried using a paragraph initially but it wasn't clear that this is a container block and thought that the '+' button makes it clearer.

> On Twenty Twenty-One, when I add several paragraphs to the Restricted Content block, the styling on the frontend does not seem to work properly.

This is caused by theme specific customizations. I have added the Group block classes to the block in cb2dc7d5f63e777aafb49c320823131a403d556b to replicate any CSS changes.

yscik commented:
"Restricted Course Content" sounds like a good name for this, since the conditions are related to the current course's status. 
Or, instead of `restricted`, it could be `conditional` or `contextual`? (Since content only shown to unenrolled visitors doesn't really align with restricted)

gkaragia commented:
I was thinking that we might use the block in other pages that's why I was against it but we solve this problem when the need arises: 21d77090c2feb5493736d7ff3819862279070c06

alexsanford commented on assets/blocks/restricted-content/index.js:
```diff
@@ -0,0 +1,75 @@
+export default {
+	title: __( 'Restricted Course Content', 'sensei-lms' ),
+	description: __(
+		'Content inside this container block can be restricted to a specified group of users. Possible options are enrolled users, not enrolled users and users which completed the course.',
```
Let's change it to "users who have completed the course".

gkaragia replied:
Fixed in 34a03073f7a505361d92fe2c8885f65abd86d3cc

alexsanford reviewed (commented):
All of the above looks good to me. I like the rename, that makes it more clear to me. The `+` button seems fine for the first release, I think. Tested on Twenty Twenty-One and Seedlet and it looks good now. Oddly, Seedlet seems to indent the content of this block, but it does the same thing for the Group block so I'd say we don't need to worry about it.

One tiny comment below re: some wording.

Otherwise, I think this is good to go for a first iteration. I didn't do any code review, do we still have any outstanding questions there? Are there any design or UX questions that we need more input on?

gkaragia commented:
> Otherwise, I think this is good to go for a first iteration. I didn't do any code review, do we still have any outstanding questions there? Are there any design or UX questions that we need more input on?

There were Peter's comments about the popover but these should be solved as with the latest rebase I switched to using the common one introduced in #3904.

alexsanford reviewed (approved):
Tested and works well, code LGTM! Note that I used the Github feature to merge the latest from `feature/lessons-block` into this branch.
